File: 03-sliding_window/00003-Longest_Substring_Without_Repeating_Characters/00003-longest_substring_without_repeating_characters.cpp
// 3: Longest Substring Without Repeating Characters
// https://leetcode.com/problems/longest-substring-without-repeating-characters/


#include <iostream>
#include <unordered_set>

using namespace std;

class Solution {
public:
    // SOLUTION
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> letters;
        int i=0, j=0;
        int result = 0;

        while (j<s.size()) {
            if (letters.find(s[j]) == letters.end()) {
                letters.insert(s[j++]);
                result = max(result, j-i);
            } else {
                letters.erase(s[i++]);
            }
        }

        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    string s = "abcabcbb";

    // OUTPUT
    auto result = o.lengthOfLongestSubstring(s);
    cout<<result<<endl;

    return 0;
}

File: 03-sliding_window/00003-Longest_Substring_Without_Repeating_Characters/00003-longest_substring_without_repeating_characters.go
// 3: Longest Substring Without Repeating Characters
// https://leetcode.com/problems/longest-substring-without-repeating-characters/

package main

import (
	"fmt"
	"math"
)

// SOLUTION
func lengthOfLongestSubstring(s string) int {
	letters := make(map[byte]int)
	i, j := 0, 0
	result := 0

	for i<len(s) {
		if _, ok := letters[s[i]]; ok {
			j = int(math.Max(float64(j), float64(letters[s[i]]+1)))
		}
		letters[s[i]] = i
		result = int(math.Max(float64(result), float64(i-j+1)))
		i++
	}

	return result
}

func main() {
	// INPUT
	s := "abcabcbb"

	// OUTPUT
	result := lengthOfLongestSubstring(s)
	fmt.Println(result)
}